include/simdjson/padded_string.h

```diff
@@ -33,77 +50,160 @@ inline char *allocate_padded_buffer(size_t length) noexcept {
+      : padded_string(sv_.data(), sv_.size()) {}
+
+  // move ctor
+  padded_string(padded_string &&other_) noexcept {
+    // calling move assignment
+    *this = std::move(other_);
   }
 
-  // note: do pass std::string_view arguments by value
-  padded_string(std::string_view sv_) noexcept
-      : viable_size(sv_.size()), data_ptr(allocate_padded_buffer(sv_.size())) {
-    if (data_ptr != nullptr) {
-      memcpy(data_ptr, sv_.data(), sv_.size());
-      data_ptr[sv_.size()] = '\0'; // easier when you need a c_str
-    }
+  // move assignment
+  padded_string &operator=(padded_string && other_ ) {
+    using namespace std;
+    std::swap(data_ptr, other_.data_ptr);
+    std::swap(viable_size, other_.viable_size);
+    // free the payload
+    simdjson::aligned_free_char(other_.data_ptr);
+    other_.data_ptr = nullptr;
+    other_.viable_size = 0;
+    return *this;
   }
 
-  padded_string(padded_string &&o) noexcept
-      : viable_size(o.viable_size), data_ptr(o.data_ptr) {
-    o.data_ptr = nullptr; // we take ownership
+  // load from file to padded_string
+  // reset the ps_
+  // return the result of fread()
+  // return 0 on error
+  static errno_t load(type  &ps_, FILE *fp) noexcept {
+
+    if (std::fseek(fp, 0, SEEK_END) < 0) {
+      std::fclose(fp);
+      return errno;
+    }
+    long llen = std::ftell(fp);
+    if ((llen < 0) || (llen == LONG_MAX)) {
+      std::fclose(fp);
+      return errno;
+    }
+
+    // now make the temporary p string
+    size_t length_ = (size_t)llen;
+
+    if (! type::reset(ps_,length_))
+      return ENOMEM;
+
+    std::rewind(fp);
+    size_t readb = std::fread(ps_.data(), 1, length_, fp);
+    std::fclose(fp);
+
+    if (readb != length_) {
+      // ps_ dtor will free on scope exit
+      return errno;
+    }
+    return 0;
   }
 
-  padded_string &operator=(padded_string &&o) {
-    data_ptr = o.data_ptr;
-    viable_size = o.viable_size;
-    o.data_ptr = nullptr; // we take ownership
-    o.viable_size = 0;
-    return *this;
+  /*
+  Makes padded_string 'Swappable'. This implementation is also ADL friendly.
+
+  using namespace simjson ;
+  padded_string a("A");
+  padded_string b("B");
+    swap(a,b) ;
+      printf("%s", a.data() ); // "B"
+  */
+  friend void swap(padded_string &left_, padded_string &right_) {
+
+    using std::swap; // bring in swap for built-in types
+
+    swap(left_.viable_size, right_.viable_size);
+    swap(left_.data_ptr, right_.data_ptr);
   }
 
-  void swap(padded_string &o) {
-    size_t tmp_viable_size = viable_size;
-    char *tmp_data_ptr = data_ptr;
-    viable_size = o.viable_size;
-    data_ptr = o.data_ptr;
-    o.data_ptr = tmp_data_ptr;
-    o.viable_size = tmp_viable_size;
+  /*
+  padded_string a("A");
+  padded_string b("B");
+       a.swap(b) ;
+         printf("%s", a.data() ); // "B"
+  */
+  void swap( padded_string &right_) {
+    std::swap(*this, right_);
   }
 
+  /*
+     padded_string a("A");
+     a.~padded_string() ;
```

Comment:
Calling the destructor directly is almost surely not idiomatic.

Reply:
hopefully, no users will do it ...